### ISSUE TITLE
tf.whereAsync throws error when run in fit() loop

### ISSUE THREAD

**Issue by burgalon** [tensorflow/tfjs#2057]
#### TensorFlow.js version
1.2.7

#### Browser version
Chrome 76

#### Describe the problem or feature request
await tf.whereAsync(nonZero) throws exception
```
Error: Argument 'x' passed to 'mean' must be a Tensor or TensorLike, but got 'Promise'
We()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:65819
mean_()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:352938
mean()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:68101
{anonymous}()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:584064
{anonymous}()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:39875
t.scopedRun()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:40013
t.tidy()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:39777
{anonymous}()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:45364
t.scopedRun()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:40013
t.gradients()@blob:https://alpha.iodide.app/05ec5021-ef09-42c0-aae1-6b5194b4b762:2:45270
```

and also promise never returns failure when being executed inside a model.fit() custom metric

#### Code to reproduce the bug / link to feature request

```
async function someTestMetric(targ, input) {
  return await tf.whereAsync(targ.gather(0,1).nonZero(0))
}
model.compile({
    optimizer: tf.train.adam(),
    loss: loss,
    metrics: metrics,
  });
model.fit(...)
```

**Comment by burgalon** [comment]
would love to try to submit a PR if somebody has a hint or direction?

**Comment by caisq** [comment]
@dsmilkov Do you expect `whereAsync` to support gradients?
